### PR TITLE
[Core] Modulize prepare input and attention metadata builder

### PR DESCRIPTION
**Pull request by comaniac**
This PR further refactors model input builder and attention metadata builder to be more modulized and maintainable. Specifically:
- Introduce an inner data class to encapsulate intermediate data.
- Put the logic of processing a particular feature (e.g., prefix caching, sliding windows, lora, mm, etc) to separate functions.
- Use pre-defined lists to apply these functions in order.
- Make `attention_matadata_builder._add_seq_group` private, and let each attention metadata builder handle by itself. This removes the ugly argument list and provides more flexibility for each attention backend to customize `add_seq_group`.

cc @Yard1 @rkooo567

### CODE REVIEW

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:273]:
```suggestion
        """Intermediate data (data in CPU before going to GPU) for the current sequence group."""
```

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:174]: Can you add big warnings that the order should not be changed? (or each API should be written in order-agnostic way and add tests? )

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:290]: not reviewing these parts assuming everything is refactoring..

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:474]:
```suggestion
            token for data in self.inter_data_list for tokens in data.input_tokens
                for token in tokens
```

this code is very confusing, but probably it is fine haha..

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:477]: can you comment what this case is?

File: vllm/worker/model_runner.py
Comment (rkooo567) [inline vllm/worker/model_runner.py:480]:
```suggestion
            position for data in self.inter_data_list for ps in data.input_positions
```

**Review by rkooo567 — APPROVED** [review]
LGTM assuming most of code is refactoring. I will also wait for @Yard1's review since he will be the one who needs to use API

File: vllm/attention/backends/flash_attn.py
Comment (Yard1) [inline vllm/attention/backends/flash_attn.py:214]: make it a property?

File: vllm/worker/model_runner.py
Comment (Yard1) [inline vllm/worker/model_runner.py:230]: why do we need to override `__init__`? that's an anti-pattern for dataclasses. we should use `__post_init__` instead or not use a dataclass in the first place.

File: vllm/worker/model_runner.py
Comment (Yard1) [inline vllm/worker/model_runner.py:174]: instead of strings, can we just have this as an object attribute and have a list of methods directly by default? that way we can avoid the `getattr`

**Review by Yard1 — COMMENTED** [review]
looks good, some nits

File: vllm/worker/model_runner.py
Comment (comaniac) [reply vllm/worker/model_runner.py:174]: Sure I'll create this list in `__init__` and directly put methods in it. I guess this won't hurt the performance.

File: vllm/worker/model_runner.py
Comment (comaniac) [reply vllm/worker/model_runner.py:230]: I did that because I wish to initialize them with `n_seqs`, but you're right I can still do that with `__post_init__`

File: vllm/worker/model_runner.py
Comment (comaniac) [reply vllm/worker/model_runner.py:174]: It's not likely to make them order-agnostic because some functions are updating/modifying fields based on existing values. I've added `WARNING` to the comment to make it more explicit.

File: vllm/worker/model_runner.py
Comment (comaniac) [reply vllm/worker/model_runner.py:474]: I can make a utility function to improve the readability.

**Comment by comaniac** [comment]
Thanks for the review and all comments should be addressed. @rkooo567 @Yard1 PTAL.

File: vllm/worker/model_runner.py
Comment (Yard1) [reply vllm/worker/model_runner.py]: I think it may even make it better because `getattr` has overhead :)

File: vllm/worker/model_runner.py
Comment (Yard1) [inline vllm/worker/model_runner.py:237]: let's make them lists instead of tuples

**Review by Yard1 — APPROVED** [review]
lgtm

File: vllm/worker/model_runner.py
Comment (comaniac) [reply vllm/worker/model_runner.py:237]: Interesting. My impression is tuple is more efficient for iteration because it's immutable, but would like to know your opinion

Comment (Yard1) [reply vllm/worker/model_runner.py:237]: making them mutable will let end user modify them easier, but it's not a big deal either way.

File: vllm/worker/model_runner.py
Comment (congcongchen123) [inline vllm/worker/model_runner.py:311]: Is it a bug? `seq_len` might be truncated according to line 300: `seq_len = min(seq_len, context_len + token_chunk_size)`

**Pull request merged**
